src/10-ipfs-0-4-0-released/index.md

```diff
@@ -0,0 +1,121 @@
+permanently into a single multiplexer that won't work for every language or
+situation, this change allows the implementations to implement the multiplexers
+of their choosing, and then negotiate which to use during the initial connection
+handshake.
+
+This modularity with stream muxing makes it easier for certain languages to
+improve performance. For example, the Go programming language may have muxado
+and yamux implementations that are really good, but many languages lack good (or
+any) implementations of them. For example, Node.js works well with spdystream in
+testing, and it would be nice to take advantage of that. And then there are
+options like 'multiplex', which may not have the same performance, but are much
+easier to implement.
+
+So by supporting as many muxers as we can, we get to choose the best
+multiplexers for the job. It also makes it much easier to implement the IPFS
+protocols in a new language. And of course, if a better multiplexers standard
+comes along, it will be easier to upgrade IPFS to support it in the future.
+
+The same code that allows us to easily select a stream muxer is also being
+used to select which ipfs sub-protocol to use for any given stream between
+peers. Now, if we need to make a breaking protocol change to any of those (like
+the DHT or bitswap) we can do so seamlessly, and provide easy backwards
+compatibility.
+
+In addition to the multiplexers changes, the protocol revamp has also improved
+efficiency and performance in a few important ways, including the elimination of
+a double wrapping of the length prefixer, and the removal of some unneccessary
+round trips between nodes.
+
+## Changes to the repo
+
+We also made a few changes to our on disk storage format (called the 'repo').
+The way object pinning (`ipfs pin add`) works has also been upgraded to be much
+more efficient, which will improve the overall speed of adding and downloading
+IPFS data. Previously, when you pinned a file recursively, it would add a
+recursive type pin to the root, and then add an indirect type pin to each child
+node. This made enumeration for GC convenient, but was very slow and made a lot
+of things needlessly complicated. We have switched to just adding the recursive
+pin to the root object, and then doing the enumeration of child nodes when we
+actually run a garbage collection.
+
+Because of this change, you will need to run a migration (from repo version 2 to
+version 3). If you update with the `ipfs-update` tool, this will be done for you
+automatically.
+
+## Other improvements, fixes
+
+In addition to a major protocol improvement and upgrade, this release adds a lot
+of new functionality, performance speedups, and stability fixes that make this
+the best version of IPFS to date.
+
+This release also includes performance and usability improvements to IPNS, the
+IPFS way to do mutable data. IPNS uses keypair encryption to allow users to
+point a pubkeyhash to an ipfs object in a way that is crytographically
```

Comment:
... cryptographically verifiable. We call this "publishing" a hash, and you can try it out using the `ipfs publish` command.